In [1]:
# Import libraries
import pandas as pd

import numpy as np
import lightgbm as lgb

# Possibly use this if we have memory issues
#import dask.dataframe as dd

from pathlib import Path

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

from surprise import Dataset
from surprise import Reader
from surprise import KNNWithMeans
from surprise import BaselineOnly
from surprise import SVDpp
from surprise import accuracy
from surprise import dump
from surprise.model_selection import train_test_split

from utilities import data_basic_utility as databasic
from utilities import dataframe_utility as dfutil
import features_utility as featutil

## File Details

Working notebook of a Full Ensemble Complete Run, this time including a full NLP Regression with LightGBM

Collaborative Filter models to use:

* KNNWithMeans
* BaselineOnly
* SVDpp

Content Filter models

* LightGBM on all non text columns
* LightGBM with Beer Context columns (non text)
* SKLearn Linear Regression on all non text columns
* LightGBM All columns including NLP


In [2]:
filePrefix = "A3_150_ensemble_v5_complete_run"
baseDataDir = "C:/Development/Data/COSC2670/Assignment3/A3data/"
subrunDir = "subruns/"
runDir = "runs/"
modelsDir = "models/"
forceRetrainModels = True

seed = databasic.get_random_seed()

In [3]:
trainFilePath = baseDataDir + 'train.tsv'
valiFilePath = baseDataDir + 'val.tsv'
featuresFilePath = baseDataDir + 'features.tsv'
testFilePath = baseDataDir + 'test.tsv'
trainFullProcessedPath = baseDataDir + 'train_features_preprocessed.csv'
valiFullProcessedPath = baseDataDir + 'vali_features_preprocessed.csv'
testFullProcessedPath = baseDataDir + 'test_features_preprocessed.csv'

# trainFilePath = baseDataDir + 'train_200k.tsv'
# valiFilePath = baseDataDir + 'vali_200k.tsv'
# featuresFilePath = baseDataDir + 'features_200k.tsv'
# testFilePath = baseDataDir + 'test_200k.tsv'
# trainFullProcessedPath = baseDataDir + 'train_features_preprocessed_200k.csv'
# valiFullProcessedPath = baseDataDir + 'vali_features_preprocessed_200k.csv'
# testFullProcessedPath = baseDataDir + 'test_features_preprocessed_200k.csv'

In [4]:
useModelLbgmBeerContext = False
useModelLbgmAllCols = False
useModelSkLinReg = False

Load the Files one by one then delete them after your done, for memory management

In [5]:
df_train = pd.read_csv(trainFilePath,sep='\t',
              names=['RowID','BeerID','ReviewerID','BeerName','BeerType','rating'])
df_train.head(10)

RowID  BeerID  ReviewerID       BeerName   BeerType  rating
0     19   12300       10635  Rauch Ür Bock  Rauchbier     4.0
1     21   12300        6547  Rauch Ür Bock  Rauchbier     4.5
2     23   12300        9789  Rauch Ür Bock  Rauchbier     4.5
3     24   12300        7372  Rauch Ür Bock  Rauchbier     5.0
4     25   12300        1302  Rauch Ür Bock  Rauchbier     4.5
5     26   12300         704  Rauch Ür Bock  Rauchbier     4.5
6     29   12300        1747  Rauch Ür Bock  Rauchbier     5.0
7     31   12300        9368  Rauch Ür Bock  Rauchbier     4.5
8     32   12300        2568  Rauch Ür Bock  Rauchbier     4.0
9     33   12300        6838  Rauch Ür Bock  Rauchbier     4.0

In [6]:
# Setup the data to be just the Reviewer and the Beer(Item) and the Rating Label we want to learn.
dfTrainFeatures = df_train.drop(['RowID','BeerName','BeerType'],axis=1)

## Collaborative Filter Models: Train

For the Collaborative Filtering Models, we only need the Training set. Train the models, then save them to file for later use

In [7]:
# Load into a Surprise dataset
reader = Reader(rating_scale=(0, 5))
dsetTrainFeatures = Dataset.load_from_df(dfTrainFeatures[['BeerID','ReviewerID', 'rating']],reader)
trainsetTrainFeatures = dsetTrainFeatures.build_full_trainset()

In [8]:
# Create each algorithm, train the model, save it to file for later, then delete the model

predictorKNN = KNNWithMeans(k=160)
featutil.trainSurpriseModel(predictorKNN, trainsetTrainFeatures, modelsDir, filePrefix, "knnwithmeans", forceRetrainModels)
del predictorKNN

predictorBaselineOnly = BaselineOnly(bsl_options = {'n_epochs': 5, 'reg_u': 3, 'reg_i': 16})
featutil.trainSurpriseModel(predictorBaselineOnly, trainsetTrainFeatures, modelsDir, filePrefix, "baselineonly", forceRetrainModels)
del predictorBaselineOnly

predictorSVDpp = SVDpp(n_factors = 10, n_epochs=20, lr_all=0.005, reg_all=0.2)
featutil.trainSurpriseModel(predictorSVDpp, trainsetTrainFeatures, modelsDir, filePrefix, "svdpp", forceRetrainModels)
del predictorSVDpp

Computing the msd similarity matrix...
Done computing similarity matrix.
The dump has been saved as file models/A3_150_ensemble_v5_complete_run_knnwithmeans_predictor.model
Estimating biases using als...
The dump has been saved as file models/A3_150_ensemble_v5_complete_run_baselineonly_predictor.model
The dump has been saved as file models/A3_150_ensemble_v5_complete_run_svdpp_predictor.model


In [9]:
# Clean up the training data used for the collaborate filters
del trainsetTrainFeatures
del reader
del dsetTrainFeatures
del dfTrainFeatures
del df_train

## Collaborative Filter Models: Predict On Validation Data 

Now we want to load the Validation set to we can predict against it and write out the subrun files, which will be used later for the Ensemble.

First, do the Predictions for the Collaborative Filter models (surprise)

In [10]:
# Read the validation data (in full)
df_vali = pd.read_csv(valiFilePath,sep='\t',
              names=['RowID','BeerID','ReviewerID','BeerName','BeerType','rating'])

reader = Reader(rating_scale=(0, 5))

idCols = ['RowID','BeerID','ReviewerID']
dfValiIds = df_vali[idCols]
dfValiFeatures = df_vali.drop(['RowID','BeerName','BeerType'],axis=1)

dsetValiFeatures = Dataset.load_from_df(dfValiFeatures[['BeerID','ReviewerID', 'rating']],reader)

In [11]:
predictValiMae_KnnWithMeans = featutil.predictSurpriseModel(modelsDir, filePrefix, "knnwithmeans", "val", dsetValiFeatures, dfValiIds, subrunDir)
predictValiMae_BaselineOnly = featutil.predictSurpriseModel(modelsDir, filePrefix, "baselineonly", "val", dsetValiFeatures, dfValiIds, subrunDir)
predictValiMae_SVDpp = featutil.predictSurpriseModel(modelsDir, filePrefix, "svdpp", "val", dsetValiFeatures, dfValiIds, subrunDir)

MAE:  0.4395
MAE for knnwithmeans: 0.43953347322741626
MAE:  0.4397
MAE for baselineonly: 0.4397473132133758
MAE:  0.4432
MAE for svdpp: 0.4432077804052646


In [12]:
# Clean up variables from the Predict Stage
del df_vali
del reader
del dfValiIds
del dfValiFeatures
del dsetValiFeatures

## Content Filter Models, train and predict

First we want to load the features and do all the data preprocessing, then we can train the different models

In [13]:
# Load the training data
df_train = pd.read_csv(trainFilePath,sep='\t',
            names=['RowID','BeerID','ReviewerID','BeerName','BeerType','rating'])

# Load the validation data. When we want to do one hot encoding, we have to do it over both datasets to ensure consistency
df_vali = pd.read_csv(valiFilePath,sep='\t',
            names=['RowID','BeerID','ReviewerID', 'BeerName','BeerType','rating'])

# Load the validation data. When we want to do one hot encoding, we have to do it over both datasets to ensure consistency
df_test = pd.read_csv(testFilePath,sep='\t',
            names=['RowID','BeerID','ReviewerID', 'BeerName','BeerType','rating'])                         

# Load the features
df_features = pd.read_csv(featuresFilePath,sep='\t',
    names=['RowID','BrewerID','ABV','DayofWeek','Month',
          'DayofMonth','Year','TimeOfDay','Gender',
          'Birthday','Text','Lemmatized','POS_Tag'])    

In [14]:
# Add the reviewer counts to each of the data sets
df_train = featutil.addReviewerReviewCount(df_train)
df_train = featutil.addBeerReviewCount(df_train)

df_vali = featutil.addReviewerReviewCount(df_vali)
df_vali = featutil.addBeerReviewCount(df_vali)

df_test = featutil.addReviewerReviewCount(df_test)
df_test = featutil.addBeerReviewCount(df_test)

In [15]:
colsToUse = ["RowID", "BrewerID", "ABV", "DayofWeek", "DayofMonth", "Month", "Year", "Gender", "TimeOfDay", "Birthday"]

df_train_data = df_train.join(df_features[colsToUse], on="RowID", how="inner", rsuffix="Feat")
df_vali_data = df_vali.join(df_features[colsToUse], on="RowID", how="inner", rsuffix="Feat")
df_test_data = df_test.join(df_features[colsToUse], on="RowID", how="inner", rsuffix="Feat")

# Remove the duplicated Row ID, also remove Beer Name at this point, we're nt using it
df_train_data = df_train_data.drop(['RowIDFeat', "BeerName"],axis=1)
df_vali_data = df_vali_data.drop(['RowIDFeat', "BeerName"],axis=1)
df_test_data = df_test_data.drop(['RowIDFeat', "BeerName"],axis=1)

In [16]:
# Clean up these dataframes now that they have been joined
del df_train
del df_vali
del df_features
del df_test

In [17]:
# do the feature transformations
df_train_data = featutil.fixNullABV(df_train_data)
df_vali_data = featutil.fixNullABV(df_vali_data)
df_test_data = featutil.fixNullABV(df_test_data)

df_train_data, df_vali_data, df_test_data = dfutil.getDummiesForTripleSets(df_train_data, df_vali_data, df_test_data, "BrewerID")

df_train_data, df_vali_data, df_test_data = dfutil.getDummiesForTripleSets(df_train_data, df_vali_data, df_test_data, "BeerType")

df_train_data, df_vali_data, df_test_data = dfutil.getDummiesForTripleSets(df_train_data, df_vali_data, df_test_data, "Gender")

df_train_data = featutil.formatDayOfWeek(df_train_data)
df_vali_data = featutil.formatDayOfWeek(df_vali_data)
df_test_data = featutil.formatDayOfWeek(df_test_data)

df_train_data = featutil.formatMonth(df_train_data)
df_vali_data = featutil.formatMonth(df_vali_data)
df_test_data = featutil.formatMonth(df_test_data)

df_train_data = featutil.formatTimeToSec(df_train_data)
df_vali_data = featutil.formatTimeToSec(df_vali_data)
df_test_data = featutil.formatTimeToSec(df_test_data)

df_train_data = featutil.convertBirthdayToAge(df_train_data)
df_vali_data = featutil.convertBirthdayToAge(df_vali_data)
df_test_data = featutil.convertBirthdayToAge(df_test_data)

(1237434, 17)


c:\Development\COSC2670\Assignment3\utilities\dataframe_utility.py:92: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df_combined.columns = df_combined.columns.str.replace(" ", "").str.replace("/", "").str.replace("-", "") \


(1237434, 2092)
(1237434, 2195)


In [18]:
print(df_train_data.shape)
print(df_vali_data.shape)
print(df_test_data.shape)
df_test_data.head()

(746207, 2196)
(243834, 2196)
(247393, 2196)


RowID  BeerID  ReviewerID  rating  ReviewerReviewCount  BeerReviewCount  \
0     18   12300       10059     NaN                  165                8   
1     20   12300        9761     NaN                  156                8   
2     30   12300        7279     NaN                  553                8   
3     46   12300        2367     NaN                  283                8   
4     47   12300        2230     NaN                   35                8   

   ABV  DayofWeek  DayofMonth  Month  ...  BeerType_SmokedBeer  \
0  7.4          7          12      6  ...                    0   
1  7.4          6          21      5  ...                    0   
2  7.4          2          12     10  ...                    0   
3  5.5          3          22      7  ...                    0   
4  5.5          2          21      7  ...                    0   

   BeerType_Tripel  BeerType_ViennaLager  BeerType_Weizenbock  \
0                0                     0                    0   
1                0                     0                    0   
2                0                     0                    0   
3                0                     0                    0   
4                0                     0                    0   

   BeerType_Wheatwine  BeerType_WinterWarmer  BeerType_Witbier  Gender_Female  \
0                   0                      0                 0              0   
1                   0                      0                 0              0   
2                   0                      0                 0              0   
3                   0                      0                 0              0   
4                   0                      0                 0              0   

   Gender_Male  Gender_unknown  
0            1               0  
1            1               0  
2            0               1  
3            1               0  
4            1               0  

[5 rows x 2196 columns]

In [19]:
# Write the test data file out so we can load it back in later so as not to have to redo this step
df_test_data.to_csv(baseDataDir + filePrefix + "_test_cleaned.csv", index=False)

del df_test_data

In [20]:
# Get all the columns
col_names = df_train_data.columns

idCols = ['RowID','BeerID','ReviewerID']
feature_cols =  col_names.drop(['RowID','BeerID','ReviewerID','rating' ])
target_col = 'rating'

# Create the sub data sets of the features and the target
dfTrainFeatures = df_train_data[feature_cols]
dfTrainTarget = df_train_data[target_col]

dfValiIds = df_vali_data[idCols]
dfValiFeatures = df_vali_data[feature_cols]
dfValiTarget = df_vali_data[target_col]


In [21]:
dfTrainFeatures.head()

ReviewerReviewCount  BeerReviewCount  ABV  DayofWeek  DayofMonth  Month  \
0                  200               23  7.4          1          23      5   
1                   10               23  7.4          1          16      5   
2                  164               23  7.4          7          10      4   
3                  432               23  7.4          3          30      3   
4                  500               23  7.4          4          24      3   

   Year  TimeOfDay  Birthday  BrewerID_1  ...  BeerType_SmokedBeer  \
0  2011      56188         0           0  ...                    0   
1  2011       1906         0           0  ...                    0   
2  2011      44246         0           0  ...                    0   
3  2011      50880         0           0  ...                    0   
4  2011      50820        37           0  ...                    0   

   BeerType_Tripel  BeerType_ViennaLager  BeerType_Weizenbock  \
0                0                     0                    0   
1                0                     0                    0   
2                0                     0                    0   
3                0                     0                    0   
4                0                     0                    0   

   BeerType_Wheatwine  BeerType_WinterWarmer  BeerType_Witbier  Gender_Female  \
0                   0                      0                 0              0   
1                   0                      0                 0              0   
2                   0                      0                 0              0   
3                   0                      0                 0              0   
4                   0                      0                 0              0   

   Gender_Male  Gender_unknown  
0            1               0  
1            1               0  
2            0               1  
3            1               0  
4            1               0  

[5 rows x 2192 columns]

In [22]:
def getFeaturesBeerContext(df1):
  consumerCols = ["DayofWeek", "DayofMonth", "Month", "TimeOfDay", "Birthday", "Gender_Male", "Gender_Female", "Gender_unknown"]
  return dfutil.getFeaturesWithoutCols(df1, consumerCols)

In [23]:
# Train the models, save them to file and then clear the model from memory
if useModelLbgmBeerContext:
    modelBeerContext = lgb.LGBMRegressor(objective="regression_l1", metric="mae", random_state=seed
        ,learning_rate=0.010443500090385492, num_leaves = 68, max_depth = 14, n_estimators = 608
    )  
    dfTrainFeatures_BeerContext = getFeaturesBeerContext(dfTrainFeatures)
    dfValiFeatures_BeerContext = getFeaturesBeerContext(dfValiFeatures)
    featutil.trainLightGbmModel(modelBeerContext, dfTrainFeatures_BeerContext, dfTrainTarget, 
        modelsDir, filePrefix, "lgbm_beercontext", forceRetrainModels)
    predictValiMae_LgbmBeerContext = featutil.predictLightGbmModel(dfValiIds, dfValiFeatures_BeerContext, dfValiTarget,
        subrunDir, modelsDir, filePrefix, "val", "lgbm_beercontext")    
    del dfTrainFeatures_BeerContext
    del dfValiFeatures_BeerContext
    del modelBeerContext

if useModelLbgmAllCols:
    modelLgbm = lgb.LGBMRegressor(objective="regression_l1", metric="mae", random_state=seed
        ,learning_rate=0.16142127923810723, num_leaves = 127, max_depth = 18, n_estimators = 811
    ) 
    featutil.trainLightGbmModel(modelLgbm, dfTrainFeatures, dfTrainTarget, 
        modelsDir, filePrefix, "lgbm_allcols", forceRetrainModels)
    predictValiMae_LgbmAllCols = featutil.predictLightGbmModel(dfValiIds, dfValiFeatures, dfValiTarget,
        subrunDir, modelsDir, filePrefix, "val", "lgbm_allcols")    
    del modelLgbm

if useModelSkLinReg:
    modelLinReg = LinearRegression()
    featutil.trainSkLinearRegModel(modelLinReg, dfTrainFeatures, dfTrainTarget, 
        modelsDir, filePrefix, "sklinearreg", forceRetrainModels)
    predictValiMae_SkLinearReg = featutil.predictSkLinearRegModel(dfValiIds, dfValiFeatures, dfValiTarget,
        subrunDir, modelsDir, filePrefix, "val", "sklinearreg")
    del modelLinReg

In [24]:
# Clean up the variables from memory
del df_train_data
del df_vali_data
del dfTrainFeatures
del dfTrainTarget
del dfValiIds
del dfValiFeatures
del dfValiTarget

## Content Filter Models with NLP

In this version, well train the Content Filter models using NLP. Assuming that the data has been preprocessed already and saved to file via the A3_130_create_full_features_processed notebook, although later, possibly add the logic into here as part of the process

In [25]:
# Load the training data
df_train = pd.read_csv(trainFullProcessedPath)

df_train.head(10)


RowID  BeerID  ReviewerID  rating  ReviewerReviewCount  BeerReviewCount  \
0     19   12300       10635     4.0                  200               23   
1     21   12300        6547     4.5                   10               23   
2     23   12300        9789     4.5                  164               23   
3     24   12300        7372     5.0                  432               23   
4     25   12300        1302     4.5                  500               23   
5     26   12300         704     4.5                  605               23   
6     29   12300        1747     5.0                  463               23   
7     31   12300        9368     4.5                   49               23   
8     32   12300        2568     4.0                  221               23   
9     33   12300        6838     4.0                  110               23   

   BeerType_Altbier  BeerType_AmericanAdjunctLager  \
0                 0                              0   
1                 0                              0   
2                 0                              0   
3                 0                              0   
4                 0                              0   
5                 0                              0   
6                 0                              0   
7                 0                              0   
8                 0                              0   
9                 0                              0   

   BeerType_AmericanAmberRedAle  BeerType_AmericanAmberRedLager  ...  \
0                             0                               0  ...   
1                             0                               0  ...   
2                             0                               0  ...   
3                             0                               0  ...   
4                             0                               0  ...   
5                             0                               0  ...   
6                             0                               0  ...   
7                             0                               0  ...   
8                             0                               0  ...   
9                             0                               0  ...   

   Lemmatized_DocVec_190  Lemmatized_DocVec_191  Lemmatized_DocVec_192  \
0              -0.046020              -0.085923              -0.020191   
1              -0.035202              -0.071197              -0.009042   
2              -0.008459              -0.038928              -0.028027   
3              -0.043639              -0.069309              -0.031036   
4              -0.021395              -0.060889              -0.019478   
5              -0.025142              -0.064408              -0.026324   
6              -0.024191              -0.071046              -0.025827   
7              -0.035988              -0.069957              -0.026050   
8              -0.031430              -0.074136              -0.017969   
9              -0.038560              -0.071023              -0.026504   

   Lemmatized_DocVec_193  Lemmatized_DocVec_194  Lemmatized_DocVec_195  \
0               0.028098               0.009554              -0.001070   
1               0.013252              -0.038783              -0.011112   
2               0.011146              -0.041214              -0.004981   
3               0.009859              -0.034390              -0.008547   
4               0.016087              -0.051270              -0.012910   
5               0.024486              -0.047075              -0.005287   
6               0.004000              -0.050098              -0.008236   
7               0.023177              -0.032557              -0.010953   
8               0.024147              -0.043018              -0.012190   
9               0.015529              -0.042741               0.001625   

   Lemmatized_DocVec_196  Lemmatized_DocVec_197  Lemmatized_DocVec_198  \
0               0.058253              -0.020965     

In [26]:
# Get all the columns
col_names = df_train.columns

idCols = ['RowID','BeerID','ReviewerID']
feature_cols =  col_names.drop(['RowID','BeerID','ReviewerID','rating' ])
target_col = 'rating'

# Create the sub data sets of the features and the target
dfTrainIds = df_train[idCols]
dfTrainFeatures = df_train[feature_cols]
dfTrainTarget = df_train[target_col]

# train the model
model_lgbm_nlp = lgb.LGBMRegressor(objective="regression_l1", metric="mae", random_state=seed
    ,learning_rate=0.09075359977364383, num_leaves = 120, max_depth = 40, n_estimators = 248, min_split_gain = 0.6310082232017945, 
    min_child_samples = 35, subsample = 0.9466694477903548, 
    subsample_freq = 0, colsample_bytree = 0.29392263338193186, reg_alpha = 0.891904482598078, reg_lambda = 0.4521335679885054
  )
featutil.trainLightGbmModel(model_lgbm_nlp, dfTrainFeatures, dfTrainTarget, 
    modelsDir, filePrefix, "lgbm_allcols_inc_nlp", forceRetrainModels)

MemoryError: Unable to allocate 14.4 GiB for an array with shape (2593, 746207) and data type float64

In [ ]:
# clean up training data from memory
del df_train
del dfTrainIds
del dfTrainFeatures
del dfTrainTarget

del model_lgbm_nlp

In [ ]:
df_vali = pd.read_csv(valiFullProcessedPath)

df_vali.head(10)

In [ ]:

dfValiIds = df_vali[idCols]
dfValiFeatures = df_vali[feature_cols]
dfValiTarget = df_vali[target_col]

predictValiMae_LgbmAllColsIncNlp = featutil.predictLightGbmModel(dfValiIds, dfValiFeatures, dfValiTarget,
    subrunDir, modelsDir, filePrefix, "val", "lgbm_allcols_inc_nlp")  
     

del df_vali
del dfValiIds
del dfValiFeatures
del dfValiTarget

## Train the Ensemble Model

Now that all the sub run files have been generated, combine all the predictions into one dataset, train a new final, ensemble model, predict on the validation data and get an MAE and save the model for use later on the Test data.

In [ ]:
# Read the validation data (in full) again. But this time, we just want the Row and the rating
df_vali = pd.read_csv(valiFilePath,sep='\t',
              names=['RowID','BeerID','ReviewerID','BeerName','BeerType','rating'])

df_ensemble_full = df_vali[["RowID", "rating"]]      

del df_vali

In [ ]:
# Load all the sub runs and join them together with the ensemble data

# Collaborative Filter Runs
fileName = filePrefix + "_" + "knnwithmeans" + "_val" + "_subrun"
df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

fileName = filePrefix + "_" + "baselineonly" + "_val" + "_subrun"
df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

fileName = filePrefix + "_" + "svdpp" + "_val" + "_subrun"
df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

# # Content Filter Runs
if useModelLbgmBeerContext:
  fileName = filePrefix + "_" + "lgbm_beercontext" + "_val" + "_subrun"
  df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

if useModelLbgmAllCols:
  fileName = filePrefix + "_" + "lgbm_allcols" + "_val" + "_subrun"
  df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

if useModelSkLinReg:
  fileName = filePrefix + "_" + "sklinearreg" + "_val" + "_subrun"
  df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)


# # Content Filter with NLP Runs
fileName = filePrefix + "_" + "lgbm_allcols_inc_nlp" + "_val" + "_subrun"
df_ensemble_full = featutil.joinRunToEnsembleFrame(df_ensemble_full, subrunDir, fileName)

In [ ]:
# Get all the columns
col_names = df_ensemble_full.columns

idCols = ['RowID']
feature_cols =  col_names.drop(['RowID','rating'])
target_col = 'rating'

# Create the sub data sets of the features and the target
dfTrainIds = df_ensemble_full[idCols]
dfTrainFeatures = df_ensemble_full[feature_cols]
dfTrainTarget = df_ensemble_full[target_col]


In [ ]:
# Doing the final Ensemble prediction using Light GBM Regression, params tuned

# Create the model and predict
model = lgb.LGBMRegressor(objective="regression_l1", metric="mae", random_state=seed,
  learning_rate=0.298864877137463, num_leaves=127, max_depth=26, n_estimators=974
)
model.fit(X=dfTrainFeatures, y=dfTrainTarget)

# use the model to predict
test_predicted = model.predict(dfTrainFeatures)
dfPredicted = pd.DataFrame({"Predict": test_predicted})

# Calc the MAE and display
predictValiMae_Ensemble = mean_absolute_error(dfTrainTarget, test_predicted)
print("Ensemble Final Average MAE (from validation data): " + str(predictValiMae_Ensemble))

# Save the model to file
model.booster_.save_model(modelsDir + filePrefix + "_ensemble_predictor.model")

print(model.feature_importances_)

In [ ]:
# Clean up all the variables
del df_ensemble_full
del dfTrainIds
del dfTrainFeatures
del dfTrainTarget
del model
del test_predicted
del dfPredicted

## Predict on the Test data with Models for Subruns

Now that we have the final Ensemble model, we can process the Test data. We need to load the test data, and create all the sub runs by using all the base level models to predict.

First, predict using the Collaborative Filter Models

In [ ]:
# Read the validation data (in full)
df_test = pd.read_csv(testFilePath,sep='\t',
              names=['RowID','BeerID','ReviewerID','BeerName','BeerType'])

# The test set is unlabeled, so we don't know the true ratings. Populate a rating col with zeros, as we are going
# to predict these values
df_test["rating"] = 0

reader = Reader(rating_scale=(0, 5))

idCols = ['RowID','BeerID','ReviewerID']
dfTestIds = df_test[idCols]
dfTestFeatures = df_test.drop(['RowID','BeerName','BeerType'],axis=1)
dsetTestFeatures = Dataset.load_from_df(dfTestFeatures[['BeerID','ReviewerID','rating']],reader)

In [ ]:
# Predict using the Collaborative Filter Models
featutil.predictSurpriseModel(modelsDir, filePrefix, "knnwithmeans", "test", dsetTestFeatures, dfTestIds, subrunDir)
featutil.predictSurpriseModel(modelsDir, filePrefix, "baselineonly", "test", dsetTestFeatures, dfTestIds, subrunDir)
featutil.predictSurpriseModel(modelsDir, filePrefix, "svdpp", "test", dsetTestFeatures, dfTestIds, subrunDir)

# Ignore the displaed MAEs, since all the targets are 0

In [ ]:
# Clean up variables from the Predict Stage
del reader
del dfTestIds
del dfTestFeatures
del dsetTestFeatures

# Keep this, as we will use this in the next stage
# del df_test 

Now Predict using the Content Filter Models. 

In [ ]:
# Reload that test data that was cleaned and processed previously
df_test_data = pd.read_csv(baseDataDir + filePrefix + "_test_cleaned.csv")

In [ ]:
# Get all the columns
col_names = df_test_data.columns

idCols = ['RowID','BeerID','ReviewerID']
feature_cols =  col_names.drop(['RowID','BeerID','ReviewerID', 'rating' ])

# Create the sub data sets of the features and the target
dfTestIds = df_test_data[idCols]
dfTestFeatures = df_test_data[feature_cols]

In [ ]:
print(df_test_data.columns)
df_test_data.head()

Problem is with one hot encoding, different sets of brewers or beer types between the training data (train+vali) and what is in test

In [ ]:
# Now we can make predictions according to each of our Content Filter Models. Pass None for the target set, the function
# will just skip the evaluation (calculating the MAE)
if useModelLbgmBeerContext:
    dfTestFeatures_BeerContext =  getFeaturesBeerContext(dfTestFeatures)
    featutil.predictLightGbmModel(dfTestIds, dfTestFeatures_BeerContext, None,
        subrunDir, modelsDir, filePrefix, "test", "lgbm_beercontext")    
    del dfTestFeatures_BeerContext

if useModelLbgmAllCols:
    featutil.predictLightGbmModel(dfTestIds, dfTestFeatures, None,
        subrunDir, modelsDir, filePrefix, "test", "lgbm_allcols") 

if useModelSkLinReg:
    featutil.predictSkLinearRegModel(dfTestIds, dfTestFeatures, None,
        subrunDir, modelsDir, filePrefix, "test", "sklinearreg")


In [ ]:
del df_test_data
del dfTestIds
del dfTestFeatures

Finally, Predict using the Content Filter with NLP models

In [ ]:
df_test = pd.read_csv(testFullProcessedPath)

df_test.head(10)

In [ ]:
# Get all the columns
col_names = df_test.columns

idCols = ['RowID','BeerID','ReviewerID']
feature_cols =  col_names.drop(['RowID','BeerID','ReviewerID', 'rating' ])

dfTestIds = df_test[idCols]
dfTestFeatures = df_test[feature_cols]

featutil.predictLightGbmModel(dfTestIds, dfTestFeatures, None,
    subrunDir, modelsDir, filePrefix, "test", "lgbm_allcols_inc_nlp")  
     

del dfTestIds
del dfTestFeatures

### Load the Ensemble Model and predict on the Test data

Load the test data

In [ ]:
df_ensemble_test = df_test[["RowID"]]      

del df_test

In [ ]:
# Load all the sub runs and join them together with the ensemble data

# Collaborative Filter Runs
fileName = filePrefix + "_" + "knnwithmeans" + "_test" + "_subrun"
df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

fileName = filePrefix + "_" + "baselineonly" + "_test" + "_subrun"
df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

fileName = filePrefix + "_" + "svdpp" + "_test" + "_subrun"
df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

# # Content Filter Runs
if useModelLbgmBeerContext:
  fileName = filePrefix + "_" + "lgbm_beercontext" + "_test" + "_subrun"
  df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

if useModelLbgmAllCols:
  fileName = filePrefix + "_" + "lgbm_allcols" + "_test" + "_subrun"
  df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

if useModelSkLinReg:
  fileName = filePrefix + "_" + "sklinearreg" + "_test" + "_subrun"
  df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)


# Content Filter Runs inc NLP doc vector cols
fileName = filePrefix + "_" + "lgbm_allcols_inc_nlp" + "_test" + "_subrun"
df_ensemble_test = featutil.joinRunToEnsembleFrame(df_ensemble_test, subrunDir, fileName)

In [ ]:
df_ensemble_test.head()

In [ ]:
# Get all the columns
col_names = df_ensemble_test.columns

idCols = ['RowID']
feature_cols =  col_names.drop(['RowID'])

# Create the sub data sets of the features and the target
dfTestFeatures = df_ensemble_test[feature_cols]

In [ ]:
# load the ensemble model  and predict
model = lgb.Booster(model_file=modelsDir + filePrefix + "_ensemble_predictor.model")
predicted = model.predict(dfTestFeatures)

dfPredictions = df_ensemble_test[idCols]
dfPredictions["Score"] = predicted

# join the predictions to the ids, sort by rowid and write to out the subrun file
finalRunFilePath = runDir + filePrefix + "_run.tsv"
dfPredictions.to_csv(finalRunFilePath, sep="\t", index=False, header=False)


In [ ]:
print("Final Report on Validation Set MAEs")
print(" ")
print("* KNN With Means: " + str(predictValiMae_KnnWithMeans))
print("* Baseline Only: " + str(predictValiMae_BaselineOnly))
print("* SVDpp: " + str(predictValiMae_SVDpp))

if useModelLbgmBeerContext:
  print("* Lgbm Beer Context columns: " + str(predictValiMae_LgbmBeerContext))

if useModelLbgmAllCols:
  print("* Lgbm All cols: " + str(predictValiMae_LgbmAllCols))

if useModelSkLinReg:
  print("* Sklearn Linear Regression: " + str(predictValiMae_SkLinearReg))
  
print("* Lgbm All cols inc NLP: " + str(predictValiMae_LgbmAllColsIncNlp))
print(" ")
print("Final Ensemble MAE: " + str(predictValiMae_Ensemble))

In [ ]:
# Clean up variables
del df_ensemble_test
del dfTestFeatures
del model
del predicted
del dfPredictions

## Run with the 200k data:

Final Report on Validation Set MAEs

* KNN With Means: 0.4378504968168548
* Baseline Only: 0.42622980784163705
* SVDpp: 0.4284517308694879
* Lgbm All cols: 0.4583928664351504
* Lgbm Beer Context columns: 0.4528618728423355
* Sklearn Linear Regression: 0.4664260836910444
* Lgbm All cols inc NLP: 0.4397404748454293

Final Ensemble MAE: 0.2865374125139253

---------

## Run with all data, all models:

Final Report on Validation Set MAEs

* KNN With Means: 0.43953347322741626
* Baseline Only: 0.4397473132133758
* SVDpp: 0.44321425908767664
* Lgbm All cols: 0.49157360211732004
* Lgbm Beer Context columns: 0.47880436516076436
* Sklearn Linear Regression: 0.49873941880294115
* Lgbm All cols inc NLP: 0.4744933270571013

Final Ensemble MAE: 0.41531043120736394

## Run on All data, no Standard Content Filter models
No lgbm All Cols, Beer Context or SKLearn Lin Reg
Only 3x Collab Filter models + LGBM All cols and NLP